### ISSUE TITLE
Propagate M1 binaries - part 2: biocpush scripts and crontab

### ISSUE THREAD

**Issue by hpages** [Bioconductor/BBS#186]
Hi Jen,

The scripts we run from the biocpush account on the central builder to effectively propagate packages are located in the [`BBS/propagation/` folder](https://github.com/Bioconductor/BBS/tree/master/propagation/). These are (in the order we run them):

- `updateReposPkgs-<buildtype>.sh`: Copies packages that received the propagation green light from `~biocbuild/public_html/BBS/X.Y/<buildtype>/OUTGOING` to the corresponding local CRAN-style repo under `~/PACKAGES/X.Y/`.

- `prepareRepos-<buildtype>.sh`: Collects all sort of information from the packages currently in the local CRAN-style repo and dumps that information in various files. The 2 most prominent kinds of files produced by `prepareRepos-<buildtype>.sh` are:
    - The PACKAGES index files. That's 3 files per sub-repo: `PACKAGES`, `PACKAGES.gz`, and `PACKAGES.rds`, i.e. that will be 12 files in total for a full CRAN-style repo where we have a new sub-repo for the Mac arm64 binaries (`bioc/bin/macosx/big-sur-arm64`).
    - The `VIEWS` file. Only one per CRAN-style repo, located in the top-level folder. Will be used downstream by the scripts that we run on staging.bioconductor.org to produce the package landing pages.

- ` pushRepos-<buildtype>.sh`: A simple call to `rsync` to re-sync the public CRAN-style repo on master.bioconductor.org with the local CRAN-style repo.

Note that only the first script "connects" the builds with the local repo. The 2 other scripts know nothing about builds: they just need to be run each time the content of the local CRAN-style repo changes. The only 2 situations when this happens is: (1) at the end of a build when we run `updateReposPkgs-<buildtype>.sh` to update the local repo with packages coming out of the builds, and (2) when some package(s) get manually added or removed from the local repo (e.g. new data-annotation packages).

Suggested roadmap:

1. For the propagation of the Mac arm64 binaries, we need a new `updateReposPkgs-<buildtype>.sh` script: `updateReposPkgs-bioc-mac-arm64.sh`. It will take care of copying the binaries that received the propagation green light from `~biocbuild/public_html/BBS/X.Y/bioc-mac-arm64/OUTGOING/mac.arm64.binary/` to the new Mac arm64 sub-repo of the local CRAN-style repo for software packages (`~/PACKAGES/X.Y/bioc`).

2. After we run `updateReposPkgs-bioc-mac-arm64.sh`, we will run `prepareRepos-bioc.sh` followed by ` pushRepos-bioc.sh`. However `prepareRepos-bioc.sh` will need to be modified to take care of the Mac arm64 sub-repo. Right now it knows nothing about this sub-repo. Note that running it as-is would not hurt, but the script would just ignore the new sub-repo. In particular the PACKAGES index files for Mac arm64 binaries would be missing or empty, so, after we run ` pushRepos-bioc.sh`, the binaries would end-up on master.bioconductor.org but people on an M1 Mac wouldn't be able to install them with `BiocManager::install()`.

3. Once `updateReposPkgs-bioc-mac-arm64.sh` and `prepareRepos-bioc.sh` are ready, we'll add a new entry in the crontab for biocpush that runs the following command:
    ```
    cd /home/biocpush/propagation/3.16 && (./updateReposPkgs-bioc-mac-arm64.sh && ./prepareRepos-bioc.sh && ./pushRepos-bioc.sh) >>/home/biocpush/cron.log/3.16/propagate-bioc-mac-arm64-`date +\%Y\%m\%d`.log 2>&1
    ```
    We will need to make sure that this doesn't run at the same time as the crontab entry for the daily builds because the 2 concurrent instances of `./prepareRepos-bioc.sh` would probably clash.

Hope this helps. When you work on 2., you'll discover that `prepareRepos-bioc.sh` performs most of its work in R by calling functionalities from the **biocViews** package. So code in this package will probably need some adjustment. If that's the case, please submit a PR for **biocViews**.

Let me know if you have questions.

Thanks,
H.

**Comment by jwokaty** [comment]
This is a note to myself that I need to change CRAN path to `bin/macosx/big-sur-arm64/` as I believe I set it to `bin/macosx/arm64`.

**Comment by hpages** [comment]
I see what you're saying, and I have a feeling that I could have done a better job at explaining why a separate `prepareRepos-bioc-big-sur-arm64.sh` is not necessary when we discussed this during our video call earlier today. So I'll try again here.

Remember that `prepareRepos-bioc.sh` is a script that we must run every time the content of our internal software CRAN-style repo changes. The script does a lot of massaging to the CRAN-style repo and it's expected to take care of _all the types of packages_. Right now it only takes care of source, Windows, and Intel Mac packages. So we need to modify it to also take care of the Mac arm64 packages.

That being said, it's true that we could probably have more specific scripts, one for each type of package, e.g.:
- `prepareRepos-bioc-source.sh`
- `prepareRepos-bioc-windows.sh`
- `prepareRepos-bioc-macosx.sh`
- `prepareRepos-bioc-big-sur-arm64.sh`

and then we would run the script that corresponds to the part of the CRAN-style repo that has changed.

However, that means more scripts to manage as well as more complicated (and thus more error prone) operating procedures. It's just easier to have _one single_ `prepareRepos-bioc.sh` script to run _whatever part of the repos was modified_. Then you don't need to know what has changed in the repo in order to decide which script to run.

Also note that the typical situation is that _every parts_ of the CRAN-style repo get updated at once. That's what happens at the end of the daily builds when we propagate packages for _all types_ (source, Windows, Mac), not just one type. Even though it's true that at the end of the Mac arm64 builds, we propagate one type of package only. But that's a temporary situation. The hope is that at some point the Mac arm64 builds will be merged to the daily builds. Then `prepareRepos-bioc.sh` will need to do the right thing anyway. So it's just easier to fix the script now and use it in the context of propagation of the arm64 binaries, even if we know that the script will do a lot of unnecessary work _in that particular context_.

Hope this helps,

As always don't hesitate to ask if you have questions.

H.

**Comment by jwokaty** [comment]
After propagation, I only find the following packages on master:

```
webadmin@ip-172-30-4-20:/extra/www/bioc/packages/3.16/bioc/bin/macosx/big-sur-arm64/contrib/4.2$ ls
GenVisR_1.29.1.tgz  PACKAGES  PACKAGES.gz  PACKAGES.rds  bandle_1.1.3.tgz  pRolocGUI_2.7.0.tgz  pRoloc_1.37.0.tgz  qPLEXanalyzer_1.15.1.tgz  waddR_1.11.0.tgz
```

But I know there should be more.

**Comment by hpages** [comment]
Well, those are the packages you have in `~biocpush/PACKAGES/3.16/bioc/bin/macosx/big-sur-arm64/contrib/4.2` right? What you see on master just reflects that.

So the real question is why do you have so few arm64 packages in the internal software repo on nebbiolo2? Is your `updateReposPkgs-bioc-mac-arm64.sh` script working? Did you run it? It should be very easy to test the script: just run it manually (it should be take a few seconds only) and look at the output. Does it look right? After it's done you can also double-check that everything looks right in `~biocpush/PACKAGES/3.16/bioc/bin/macosx/big-sur-arm64/contrib/4.2`.

Only after you had a successful run of `updateReposPkgs-bioc-mac-arm64.sh` (step 1 in Suggested roadmap above) you should work on `prepareRepos-bioc.sh` (step 2 in Suggested roadmap above).

H.

**Comment by hpages** [comment]
Actually it seems that `updateReposPkgs-bioc-mac-arm64.sh` just obeyed the propagation LEDs that are displayed on the Mac ARM64 report: https://bioconductor.org/checkResults/3.16/bioc-mac-arm64-LATEST/

Not sure why only 6 LEDs are green. Looks like something went wrong here (extracted from `nebbiolo2-20220914-postrun.log`):
```
BBS> ==============================================================
BBS> [stage6c] STARTING stage6c on Wed Sep 14 17:00:32 2022...
  Rscript> START creating PROPAGATION_STATUS_DB.txt ...
  Rscript> - write_PACKAGES() to OUTGOING/mac.arm64.binary/ ... OK
  Rscript> - Start computing propagation statuses for "mac.arm64.binary" packages:
  Rscript>   - nb of candidates = 25 (based on version > published version)
  Rscript>   - pass #1 ... OK ==> total approved candidates = 4
  Rscript>   - pass #2 ... OK ==> total approved candidates = 6
  Rscript>   - pass #3 ... OK ==> total approved candidates = 6
  Rscript> - Done computing propagation statuses for "mac.arm64.binary" packages.
  Rscript> - Write computed statuses to PROPAGATION_STATUS_DB.txt ... OK
  Rscript> DONE creating PROPAGATION_STATUS_DB.txt.
BBS> [stage6c] DONE on Wed Sep 14 17:02:37 2022.
```
Total number of approved candidates for propagation is 6!

I remember that you had a great report not so long ago with plenty of green LEDs. We can still this in the logs:
```
biocbuild@nebbiolo2:~/bbs-3.16-bioc-mac-arm64/log$ grep 'total approved candidates' *.log
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #1 ... OK ==> total approved candidates = 333
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #2 ... OK ==> total approved candidates = 448
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #3 ... OK ==> total approved candidates = 618
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #4 ... OK ==> total approved candidates = 724
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #5 ... OK ==> total approved candidates = 765
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #6 ... OK ==> total approved candidates = 885
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #7 ... OK ==> total approved candidates = 1055
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #8 ... OK ==> total approved candidates = 1214
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #9 ... OK ==> total approved candidates = 1373
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #10 ... OK ==> total approved candidates = 1548
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #11 ... OK ==> total approved candidates = 1723
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #12 ... OK ==> total approved candidates = 1804
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #13 ... OK ==> total approved candidates = 1856
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #14 ... OK ==> total approved candidates = 1881
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #15 ... OK ==> total approved candidates = 1885
nebbiolo2-20220824-postrun.log:  Rscript>   - pass #16 ... OK ==> total approved candidates = 1885
nebbiolo2-20220904-postrun.log:  Rscript>   - pass #1 ... OK ==> total approved candidates = 1
nebbiolo2-20220904-postrun.log:  Rscript>   - pass #2 ... OK ==> total approved candidates = 3
nebbiolo2-20220904-postrun.log:  Rscript>   - pass #3 ... OK ==> total approved candidates = 3
nebbiolo2-20220907-postrun.log:  Rscript>   - pass #1 ... OK ==> total approved candidates = 3
nebbiolo2-20220907-postrun.log:  Rscript>   - pass #2 ... OK ==> total approved candidates = 5
nebbiolo2-20220907-postrun.log:  Rscript>   - pass #3 ... OK ==> total approved candidates = 5
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #1 ... OK ==> total approved candidates = 4
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #2 ... OK ==> total approved candidates = 6
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #3 ... OK ==> total approved candidates = 6
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #1 ... OK ==> total approved candidates = 4
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #2 ... OK ==> total approved candidates = 6
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #3 ... OK ==> total approved candidates = 6
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #1 ... OK ==> total approved candidates = 4
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #2 ... OK ==> total approved candidates = 6
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #3 ... OK ==> total approved candidates = 6
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #1 ... OK ==> total approved candidates = 4
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #2 ... OK ==> total approved candidates = 6
nebbiolo2-20220914-postrun.log:  Rscript>   - pass #3 ... OK ==> total approved candidates = 6
``` 
Don't know what happened on September 4, but you want to go back to a situation similar to what we had before that.

Once you have that (and the correct propagation status in `~/public_html/BBS/3.16/bioc-mac-arm64/PROPAGATION_STATUS_DB.txt`), make sure to completely freeze any activity on the `biocbuild` build side of the Mac ARM64 builds (i.e. disable `prerun.sh` and `postrun.sh`, and also disable `run.sh` on kjohnson), so you keep the good report for all the time that you need to work on the `biocpush` side of the propagation pipe.

**Comment by jwokaty** [comment]
I think I need to rename the binaries to `big-sur-arm64` since that's the type that produces the correct contrib.url. See https://github.com/Bioconductor/BiocManager/issues/139

**Comment by hpages** [comment]
The name we use internally (`mac.arm64.binary`) doesn't have much to do with that. What's important is that we use the same _layout_ as the CRAN people for our CRAN-style repo i.e. we must put our arm64 binaries in `bioc/bin/macosx/big-sur-arm64` like they do. I think that's what you're doing so all is fine.

You also want to make sure that you supply the right `type` value when you call `write_PACKAGES()` to produce the `PACKAGES` index file in `bioc/bin/macosx/big-sur-arm64`. I'm not sure `type="mac.binary.big-sur-arm64"` is supported so it seems that you have to use `type="mac.binary"` for that one.

**Comment by jwokaty** [comment]
The layout is correct. I think I need to do this for `contrib.url` since it can already handle 'mac.binary.big-sur-arm64' but not 'mac.arm64.binary', where something is between 'mac' and 'binary'. I think it will be less confusing if I do it this way and I don't need to do extra work to use `contrib.url`.

I currently pass 'mac.binary' to `write_PACKAGES()`; however, since that didn't update the `PACKAGES` files correctly, I might need to do some extra work. I have to look.

**Comment by hpages** [comment]
It's not a problem that `contrib.url()` cannot handle `"mac.arm64.binary"`. `mac.arm64.binary` is for our internal use only (i.e. in BBS/biocViews code). It's not meant to be used directly as a valid `type` value for `contrib.url()` or `write_PACKAGES()`.

I mean, you can replace `mac.arm64.binary` with `mac.binary.big-sur-arm64` in BBS/biocViews code, but you'll still need to do some "translation" e.g. when you call `write_PACKAGES()`. So it seems that whatever internal name we use, it won't be a valid `type` value everywhere, unfortunately.

**Comment by hpages** [comment]
FWIW, the `write_PACKAGES()` issue has been visible in the `~biocpush/cron.log/3.16/propagate-bioc-202209*.log` files for many days (and also in the `propagate-bioc-mac-arm-64-20220921.log` file from yesterday).

**Comment by hpages** [comment]
Could the problem be that you didn't deploy your latest changes to **biocViews**? Latest version is 1.65.6 but I still see 1.65.5 on nebbiolo2 (for R-4.2 in the biocpush account).

**Comment by jwokaty** [comment]
Yeah, I should have caught the `write_PACKAGES` problem. I have to write a similar function to solve the problem.

**Comment by hpages** [comment]
> I have to write a similar function to solve the problem.

Deploying your latest changes to **biocViews** didn't solve it?

**Comment by hpages** [comment]
Just want to confirm that `tools::write_PACKAGES(..., type="mac.binary")` seems to work fine on `~/PACKAGES/3.16/bioc/bin/macosx/big-sur-arm64/contrib/4.2/` (I just tried it), so there should be no need to write a similar function for the arm64 binaries.

The problem I see is that **biocViews** 1.65.5 was not calling `write_PACKAGES()` properly but I believe that you corrected that bug in 1.65.6 so all you need to do is use **biocViews** 1.65.6. Maybe there are other problems but I wouldn't try to look too far without re-running `prepareRepos-bioc.sh` with the latest **biocViews** first.

**Comment by jwokaty** [comment]
Thanks. `tools::write_PACKAGES` does work the `arm64` path. I was confused as to why it wasn't being updated, but I needed to add a `gsub` line to transform the type to `mac.binary`. It seems working correctly now.

**Comment by hpages** [comment]
Great. Thanks for sorting this out.

Note that `pushRepos-bioc-mac-arm64.sh` is probably not needed as it's just a copy of `pushRepos-bioc.sh`.

We'll also need to work on adding the new arm64-related info to the `VIEWS` file. This will require more changes to the **biocViews** package I think. Let's discuss next week. I'm going to create a new issue for that.
